### PR TITLE
Azure rm subscription info

### PR DESCRIPTION
**Pull request by paultaiton**
Create module for azure_rm_subscription_info

##### SUMMARY

There does not exist an easy way to fetch Azure Subscription information from within Ansible. This PR creates a new azure_rm_subscription_info module that allows for querying of subscriptions based on id, name, or to fetch a list of all subscriptions.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
azure_rm_subscription_info

##### ADDITIONAL INFORMATION

This new module being strictly informational is straightforward. However I had to add some conditional logic in azure_rm_common to suppress the passing of "subscription_id" to the SubscriptionClient instantiation. I tried to do it in a client agnostic manner, but I'm not aware of any others that also require it.

I have included some simplistic integration tests, however one of the files, 'aliases', I am unfamiliar with and will need some assistance with to validate that the contents that I copied from another is sufficient.

### CODE REVIEW

File: plugins/module_utils/azure_rm_common.py [hunk 1]
```diff
@@ -955,6 +961,20 @@ def storage_client(self):
...
         return StorageManagementClient.models("2019-06-01")
 
+    @property
+    def subscription_client(self):
+        self.log('Getting subscription client...')
+        if not self._subscription_client:
+            self._subscription_client = self.get_mgmt_svc_client(SubscriptionClient,
+                                                            base_url=self._cloud_environment.endpoints.resource_manager,
+                                                            suppress_subscription_id=True,
+                                                            api_version='2019-11-01')
```
Comment (Fred-sun) [inline plugins/module_utils/azure_rm_common.py]: 969，970 and 971 lines continuation line under-indented for visual indent。

File: plugins/modules/azure_rm_subscription_info.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        self.tags = None
+        self.all = False
+
+        mutually_exclusive = [['name', 'id']]
+
+        super(AzureRMSubscriptionInfo, self).__init__(self.module_arg_spec,
+                                                        supports_tags=False,
+                                                        mutually_exclusive=mutually_exclusive,
+                                                        facts_module=True)
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_subscription_info.py:158]: 155,156 and 157 lines continuation line over-indented for visual indent

File: plugins/modules/azure_rm_subscription_info.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        try:
+            response = self.subscription_client.subscriptions.list()
+        except CloudError as exc:
+            self.fail("Failed to list all items - {0}".format(str(exc)))
+
+        results = []
+        for item in response:
+            # If the name matches, return result regardless of anything else.
+            # If name is not defined and either state is Enabled or all is true, return result.
+            if ( self.name and self.name.lower() == item.display_name.lower() ):
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_subscription_info.py]: whitespace after '(' and ')'

File: plugins/modules/azure_rm_subscription_info.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                results.append(self.to_dict(item) )
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_subscription_info.py]: whitespace after '(' and ')'

File: plugins/modules/azure_rm_subscription_info.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            elif ( not self.name and ( self.all or item.state == "Enabled" ) and self.has_tags(item.tags, self.tags) ):
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_subscription_info.py]: whitespace after '(' and ')'

File: plugins/modules/azure_rm_subscription_info.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+                results.append(self.to_dict(item) )
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_subscription_info.py]: whitespace after '(' and ')'

File: plugins/modules/azure_rm_subscription_info.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+    def to_dict(self, subscription_object):
+        return dict(
+            display_name=subscription_object.display_name,
+            fqid=subscription_object.id,
+            state=subscription_object.state,
+            subscription_id=subscription_object.subscription_id,
+            tags=subscription_object.tags,
+            tenant_id=subscription_object.tenant_id
+        )
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_subscription_info.py:217]: expected 2 blank lines, found 1

File: plugins/modules/azure_rm_subscription_info.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+def main():
+    AzureRMSubscriptionInfo()
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_subscription_info.py:221]: expected 2 blank lines after class or function definition, found 1

File: plugins/modules/azure_rm_subscription_info.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+        description:
+            - Limit results by providing a list of tags. Format tags as 'key' or 'key:value'.
+            - Option has no effect when searching by id or name, and will be silently ignored.
+        type: list
+
+extends_documentation_fragment:
+    - azure.azcollection.azure
+
+author:
+    - Paul Aiton ( @paultaiton )
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_subscription_info.py]:
```suggestion
    - Paul Aiton (@paultaiton)
```

File: plugins/modules/azure_rm_subscription_info.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+  azure_rm_subscription_info:
+    name: "my-subscription"
+
+- name: Get facts for all subscriptions, including ones that are disabled.
+  azure_rm_subscription_info:
+    all: True
+
+- name: Get facts for subscriptions containing tags provided.
+    azure_rm_subscription_info:
+    tags:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_subscription_info.py:75]: The indentation error

File: plugins/modules/azure_rm_subscription_info.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+
+AZURE_OBJECT_CLASS = 'Subscription'
+
+
+class AzureRMSubscriptionInfo(AzureRMModuleBase):
+
+    def __init__(self):
+
+        self.module_arg_spec = dict(
+            name=dict(type='str'),
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_subscription_info.py]:
```suggestion
            name=dict(type='str', aliases=['subscription_name']),
```

File: plugins/modules/azure_rm_subscription_info.py [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+            id=dict(type='str'),
+            tags=dict(type='list'),
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_subscription_info.py]:
```suggestion
            tags=dict(type='list',  elements='dict'),
```

**Review by Fred-sun — COMMENTED** [review]
Please add the next two lines to the tests/sanity/ignore-2.10.txt file
```e.g.
plugins/modules/azure_rm_subscription_info.py validate-modules:required_if-requirements-unknown
plugins/modules/azure_rm_subscription_info.py validate-modules:required_if-unknown-key

File: tests/integration/targets/azure_rm_subscription/tasks/main.yml [hunk 13]
```diff
@@ -0,0 +1,24 @@
+- name: Get list of all subscriptions
+  azure.azcollection.azure_rm_subscription_info:
```
Comment (Fred-sun) [inline tests/integration/targets/azure_rm_subscription/tasks/main.yml]: Everything below needs to be changed like this!

```suggestion
  azure_rm_subscription_info:
```

**Comment by Fred-sun** [comment]
Please modify pr-pipelines.yml file , add 'azure_rm_subscription'.

File: plugins/modules/azure_rm_subscription_info.py (hunk 2 above)
Comment (paultaiton) [reply plugins/modules/azure_rm_subscription_info.py:158]: I think I've got it.

File: plugins/modules/azure_rm_subscription_info.py (hunk 3 above)
Comment (paultaiton) [reply plugins/modules/azure_rm_subscription_info.py]: I think I've corrected this, however there's quite a few in there so another look over would be appreciated.

**Comment by paultaiton** [comment]
> Please modify pr-pipelines.yml file , add 'azure_rm_subscription'.

I have added this entry.

File: plugins/modules/azure_rm_subscription_info.py (hunk 10 above)
Comment (paultaiton) [reply plugins/modules/azure_rm_subscription_info.py:75]: This has been fixed.

**Comment by paultaiton** [comment]
@Fred-sun 
Appreciate your review. I believe I have implemented all suggestions.

Thanks.

File: tests/sanity/ignore-2.10.txt [hunk 14]
```diff
@@ -276,6 +276,8 @@ plugins/modules/azure_rm_subnet.py validate-modules:required_if-unknown-key
 plugins/modules/azure_rm_subnet_info.py validate-modules:parameter-type-not-in-doc
 plugins/modules/azure_rm_subnet_info.py validate-modules:required_if-requirements-unknown
 plugins/modules/azure_rm_subnet_info.py validate-modules:required_if-unknown-key
+plugins/modules/azure_rm_subscription_info.py validate-modules:required_if-requirements-unknown
+plugins/modules/azure_rm_subscription_info.py validate-modules:required_if-unknown-key
```
Comment (Fred-sun) [inline tests/sanity/ignore-2.10.txt:280]: Add new line
```e.g.
plugins/modules/azure_rm_subscription_info.py validate-modules:doc-elements-mismatch

File: plugins/modules/azure_rm_subscription_info.py [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+
+
+class AzureRMSubscriptionInfo(AzureRMModuleBase):
+
+    def __init__(self):
+
+        self.module_arg_spec = dict(
+            name=dict(type='str', aliases=['subscription_name']),
+            id=dict(type='str'),
+            tags=dict(type='list',  elements='dict'),
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_subscription_info.py]:
```suggestion
            tags=dict(type='list', elements='dict'),
```

**Review by Fred-sun — COMMENTED** [review]
Delete the spaces after 201, 202,203 and 204 lines!
Delete the space before 201,202,203 and 204 lines!

**Comment by paultaiton** [comment]
> Delete the spaces after 201, 202,203 and 204 lines!
> Delete the space before 201,202,203 and 204 lines!

I do not fully understanding this request. I've removed the wrapping parenthesis since they are no longer necessary, and formatted it to what I think you want, but I'm not sure to what specific thing you're referring.

**Comment by Fred-sun** [comment]
> > Delete the spaces after 201, 202,203 and 204 lines!
> > Delete the space before 201,202,203 and 204 lines!
> 
> I do not fully understanding this request. I've removed the wrapping parenthesis since they are no longer necessary, and formatted it to what I think you want, but I'm not sure to what specific thing you're referring.

In order to comply with the python code pep8 specification! Thank you!

**Comment by Fred-sun** [comment]
Add the following three lines to tests/sanity/ignore-2.11.txt
```e.g.
plugins/modules/azure_rm_subscription_info.py validate-modules:required_if-requirements-unknown
plugins/modules/azure_rm_subscription_info.py validate-modules:required_if-unknown-key
plugins/modules/azure_rm_subscription_info.py validate-modules:doc-elements-mismatch

**Comment by paultaiton** [comment]
@Fred-sun 

> Add the following three lines to tests/sanity/ignore-2.11.txt

This has been done.

**Pull request merged**